core/dispatch.c

```diff
@@ -1803,20 +1806,23 @@ adjust_syscall_continuation(dcontext_t *dcontext)
                 dcontext->asynch_target);
         }
 #    endif
-    } else if (vsyscall_syscall_end_pc != NULL &&
-               /* PR 341469: 32-bit apps (LOL64) on AMD hardware have
-                * OP_syscall in a vsyscall page
-                */
-               get_syscall_method() != SYSCALL_METHOD_SYSCALL) {
+    } else if (vsyscall_syscall_end_pc != NULL) {
+        /* PR 341469: 32-bit apps (LOL64) on AMD hardware have
+         * OP_syscall in a vsyscall page
+         */
+
         /* If we fail to hook we currently bail out to int; but we then
          * need to manually jump to the sysenter return point.
          * Once we have PR 288330 we can remove this.
          */
-        if (dcontext->asynch_target == vsyscall_syscall_end_pc) {
-            ASSERT(vsyscall_sysenter_return_pc != NULL);
-            dcontext->asynch_target = vsyscall_sysenter_return_pc;
-            LOG(THREAD, LOG_SYSCALLS, 3, "%s: asynch_target => " PFX "\n", __FUNCTION__,
-                dcontext->asynch_target);
+        if (IF_X86_32((syscall_method_is_syscall && cpu_info.vendor == VENDOR_AMD) ||)
```

Comment:
I don't follow: the previous code knew about AMD (see the comment above) and deliberately excluded SYSCALL -- but now this PR adds logic to include SYSCALL for AMD which is the opposite.  If the original was wrong: why keep the `!syscall_method_is_syscall` at all?

If you could elaborate on what the problem is and how this solves it -- is vsyscall_sysenter_return_pc even set properly for SYSCALL (or is it only set for SYSENTER)?

Reply:
The problem is in hooking.

DR hooks **__kernel_vsyscall** method, but in the case when it contains `sysenter` instruction. 
It is correct on Intel machines. But on my AMD machine ___kernel_vsyscall_ contains `syscall` instead of `sysenter` and DR do not hook it.


```
____AMD____
  <__kernel_vsyscall>          push   %ecx
  <__kernel_vsyscall+1>        push   %edx
  <__kernel_vsyscall+2>        push   %ebp 
  <__kernel_vsyscall+3>        mov    %ecx,%ebp
  <__kernel_vsyscall+5>        syscall
  <__kernel_vsyscall+7>        int    $0x80
  <__kernel_vsyscall+9>        pop    %ebp
  <__kernel_vsyscall+10>       pop    %edx
  <__kernel_vsyscall+11>       pop    %ecx
  <__kernel_vsyscall+12>       ret


____ Intel_____
  <__kernel_vsyscall>          push   %ecx
  <__kernel_vsyscall+1>        push   %edx
  <__kernel_vsyscall+2>        push   %ebp 
  <__kernel_vsyscall+3>        mov    %esp,%ebp
  <__kernel_vsyscall+5>        sysenter
  <__kernel_vsyscall+7>        int    $0x80
  <__kernel_vsyscall+9>        pop    %ebp
  <__kernel_vsyscall+10>       pop    %edx
  <__kernel_vsyscall+11>       pop    %ecx
  <__kernel_vsyscall+12>       ret

```

https://github.com/DynamoRIO/dynamorio/blob/3d775872c261c11d4f9ca1d621981301cd929ceb/core/arch/arch.c#L3257-L3266

Hooking occurs when DR checks system call instructions. hook_vsyscall() method is under condition and can be executed only when OP is `sysenter`.
I think it is wrong and DR should hooks vsyscall page when run on AMD and OP is syscall.

Reply:
I suggest using logs and the debugger to understand what is going wrong, and then we can design and understand an appropriate fix.

FTR here is what happened in #5085:
```
Breakpoint 2, adjust_syscall_continuation (dcontext=0x4036b040) at /home/bruening/dr/git/src/core/dispatch.c:1817
1817	            dcontext->asynch_target = vsyscall_sysenter_return_pc;
(gdb) bt
#0  adjust_syscall_continuation (dcontext=0x4036b040) at /home/bruening/dr/git/src/core/dispatch.c:1817
#1  0xf7cb3370 in handle_post_system_call (dcontext=0x4036b040) at /home/bruening/dr/git/src/core/dispatch.c:2203
#2  0xf7caa782 in dispatch_enter_dynamorio (dcontext=0x4036b040) at /home/bruening/dr/git/src/core/dispatch.c:892
#3  0xf7ca63a8 in d_r_dispatch (dcontext=0x4036b040) at /home/bruening/dr/git/src/core/dispatch.c:160
(gdb) p get_syscall_method()
$1 = 1
(gdb) x/4i dcontext->next_tag
   0x403b5240:	int    $0x80
   0x403b5242:	mov    %eax,0x4036b05c
   0x403b5247:	mov    $0xf7ecd42c,%eax
   0x403b524c:	jmp    0x403b4f40
(gdb) x/4i dcontext->asynch_target
   0xf7c0e547:	int    $0x80
   0xf7c0e549:	pop    %ebp
   0xf7c0e54a:	pop    %edx
   0xf7c0e54b:	pop    %ecx
(gdb) p vsyscall_sysenter_return_pc
$3 = (app_pc) 0xf7c0e549 "]ZYÐ\220\220\220\215\264&"
(gdb) p vsyscall_syscall_end_pc
$4 = (app_pc) 0xf7c0e547 "̀]ZYÐ\220\220\220\215\264&"
(gdb) p vsyscall_sysenter_displaced_pc
$12 = (app_pc) 0x0
```

The vsyscall code in vdso:
```
   0xf7fcf540 <__kernel_vsyscall>:	push   %ecx
   0xf7fcf541 <__kernel_vsyscall+1>:	push   %edx
   0xf7fcf542 <__kernel_vsyscall+2>:	push   %ebp
   0xf7fcf543 <__kernel_vsyscall+3>:	mov    %esp,%ebp
   0xf7fcf545 <__kernel_vsyscall+5>:	sysenter 
   0xf7fcf547 <__kernel_vsyscall+7>:	int    $0x80
   0xf7fcf549 <__kernel_vsyscall+9>:	pop    %ebp
   0xf7fcf54a <__kernel_vsyscall+10>:	pop    %edx
   0xf7fcf54b <__kernel_vsyscall+11>:	pop    %ecx
   0xf7fcf54c <__kernel_vsyscall+12>:	ret    
```

So it has the `int` in the slot after the `sysenter`, and `adjust_syscall_continuation` is needed to skip over that for the sysenter case where the hook failed.

Reply:
Instead of using `sysenter` AMD architecture uses `syscall` (on LOL64 mode).


> The problem is in hooking.
> 
> DR hooks **__kernel_vsyscall** method, but in the case when it contains `sysenter` instruction.
> It is correct on Intel machines. But on my AMD machine ___kernel_vsyscall_ contains `syscall` instead of `sysenter` and DR do not hook it.
> 
> ```
> ____AMD____
>   <__kernel_vsyscall>          push   %ecx
>   <__kernel_vsyscall+1>        push   %edx
>   <__kernel_vsyscall+2>        push   %ebp 
>   <__kernel_vsyscall+3>        mov    %ecx,%ebp
>   <__kernel_vsyscall+5>        syscall
>   <__kernel_vsyscall+7>        int    $0x80
>   <__kernel_vsyscall+9>        pop    %ebp
>   <__kernel_vsyscall+10>       pop    %edx
>   <__kernel_vsyscall+11>       pop    %ecx
>   <__kernel_vsyscall+12>       ret
> 
> 
> ____ Intel_____
>   <__kernel_vsyscall>          push   %ecx
>   <__kernel_vsyscall+1>        push   %edx
>   <__kernel_vsyscall+2>        push   %ebp 
>   <__kernel_vsyscall+3>        mov    %esp,%ebp
>   <__kernel_vsyscall+5>        sysenter
>   <__kernel_vsyscall+7>        int    $0x80
>   <__kernel_vsyscall+9>        pop    %ebp
>   <__kernel_vsyscall+10>       pop    %edx
>   <__kernel_vsyscall+11>       pop    %ecx
>   <__kernel_vsyscall+12>       ret
> ```
> 
> 




Here is the result.

```
Breakpoint 1, adjust_syscall_continuation (dcontext=0x412ac040) at /root/dynamorio/dynamorio/core/dispatch.c:1784
1784	    bool syscall_method_is_syscall = get_syscall_method() == SYSCALL_METHOD_SYSCALL;
(gdb) disass 0xf7c11060, +20
Dump of assembler code from 0xf7c11060 to 0xf7c11074:
   0xf7c11060:	push   %ecx
   0xf7c11061:	push   %edx
   0xf7c11062:	push   %ebp
   0xf7c11063:	mov    %ecx,%ebp
   0xf7c11065:	syscall 
   0xf7c11067:	int    $0x80
   0xf7c11069:	jmp    0x4127fa42
   0xf7c1106e:	pop    %ebp
   0xf7c1106f:	pop    %edx
   0xf7c11070:	pop    %ecx
   0xf7c11071:	ret    
   0xf7c11072:	mov    $0x26,%ah
End of assembler dump.
(gdb) bt
#0  adjust_syscall_continuation (dcontext=0x412ac040) at /root/dynamorio/dynamorio/core/dispatch.c:1820
#1  0xf7cb596f in handle_post_system_call (dcontext=0x412ac040) at /root/dynamorio/dynamorio/core/dispatch.c:2209
#2  0xf7cacd55 in dispatch_enter_dynamorio (dcontext=0x412ac040) at /root/dynamorio/dynamorio/core/dispatch.c:892
#3  0xf7ca8955 in d_r_dispatch (dcontext=0x412ac040) at /root/dynamorio/dynamorio/core/dispatch.c:160
#4  0x4127efce in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb) x/4i dcontext->next_tag
   0x412f6240:	syscall 
   0x412f6242:	mov    %eax,0x412ac05c
   0x412f6247:	mov    $0xf7ecebec,%eax
   0x412f624c:	jmp    0x412f5f40
(gdb) x/4i dcontext->asynch_target
   0xf7c11067:	int    $0x80
   0xf7c11069:	jmp    0x4127fa42
   0xf7c1106e:	pop    %ebp
   0xf7c1106f:	pop    %edx
(gdb) p vsyscall_sysenter_return_pc
$4 = (app_pc) 0xf7c11069 "\351\324\351fI]ZYô&"
(gdb) p vsyscall_syscall_end_pc
$5 = (app_pc) 0xf7c11067 "̀\351\324\351fI]ZYô&"
(gdb) p vsyscall_sysenter_displaced_pc
$6 = (app_pc) 0xf7c1106e "]ZYô&"

```

In my case PC is not 0x0, but points after `jmp`,  Have you some idea?

Reply:
The vsyscall sequence doesn't make sense to me: how does it work natively (w/o DR)?
```
   0xf7c11065:	syscall 
   0xf7c11067:	int    $0x80
```
`syscall` stores the PC right afterward in `rcx` and `sysret` jumps there: so it would execute `syscall` and then execute `int 0x80` right afterward which can't be right.

Also, what is that `jmp    0x4127fa42`?  What is at that target address?  That also looks suspect.  What does this sequence look like natively?

Reply:
```
  <__kernel_vsyscall>          push   %ecx
  <__kernel_vsyscall+1>        push   %edx
  <__kernel_vsyscall+2>        push   %ebp 
  <__kernel_vsyscall+3>        mov    %ecx,%ebp
  <__kernel_vsyscall+5>        syscall
  <__kernel_vsyscall+7>        int    $0x80
  <__kernel_vsyscall+9>        pop    %ebp
  <__kernel_vsyscall+10>       pop    %edx
  <__kernel_vsyscall+11>       pop    %ecx
  <__kernel_vsyscall+12>       ret
```

This is the original body of `__kernel_vsyscall` (without DR). After `syscall` execute `int $0x80` and my program works fine. So I think it is right.

`jmp 0x4127fa42` instruction inserted in hook_vsyscall() method and target address calculated through **after_do_shared_syscall_addr()**. From that address make jump to code generated by **append_fcache_return_common()**

Reply:
Please describe the meaning of mangling in DR, or give some link. The code generated after mangling is different from code in debbuger.

Reply:
> After syscall execute int $0x80 and my program works fine. So I think it is right.

But why would it execute two system call gateways for one request?  That does not make sense to me.  Under the debugger where does control go after the `syscall`?  I find it hard to believe it would then go back into the kernel again.

Reply:
I would guess that the kernel adds 2 to `ecx` prior to the `sysret` and it comes back *after* the `int` or something like that.

Reply:
>  Under the debugger where does control go after the syscall


I set breakpoint on `int` and `pop` instructions. Debugger stopped on `pop`.
![Screenshot from 2021-09-15 00-07-43](https://user-images.githubusercontent.com/18750457/133326634-a2ca2acc-8228-4d10-9c02-0fcb05334556.png)

Reply:
> But I thought DR would not try to insert a hook unless the syscall method was sysenter?

You are right, DR on master branch does not insert hook for `syscall` instraction. I added that feature, only on AMD and LOL64 DR insert hook.

Maybe we little bit misunderstood each other :-)

Reply:
> I would guess that the kernel adds 2 to ecx prior to the sysret and it comes back after the int or something like that.

I agree with you, but what is meaning to generate code which will not be executed at all?

Reply:
Given this behavior where `syscall` does *not* return to the subsequent PC, we need to understand:

+ Does this only happen for the `syscall` in the VDSO or other `syscall` instances?  Are there instances of `syscall` inlined into libc or elsewhere?  What do they do?
+ What happens with DR generating its own `syscall` instructions?  Does the kernel add 2 to the PC, or does it set the PC to the absolute PC of the VDSO post-`syscall`-post-`int` location?

If the kernel is setting the absolute return PC of every `syscall` for 32-bit AMD, then it's behaving like `sysenter` and indeed DR needs treat it like `sysenter`: it needs to hook the vsyscall code and it has to be prepared for control to jump after any `syscall` it generates.

Reply:
I wrote assembler code with `syscall` instruction after execution specified system function kernel went to *__kernel_vsyscall* and my program crashed. 

The same behavior in DR. DR inserts `syscall` instruction through _handle_system_call()_ method. After execution `syscall` kernel goes to *__kernel_vsyscall* and cannot go back for dispatch and DR crushes.
Through hook, I add `jmp` to the instruction after `syscall` inserted in handle time, which guarantees the correct routine.

Reply:
Thank you for investigating so we can better understand this.  Given this behavior, where the kernel turns `syscall` from returning to the subsequent PC into an abnormal control flow sequence like `sysenter`, then I agree that DR needs to hook the vsyscall sequence and needs to treat `syscall` like `sysenter` for this AMD 32-bit case.

Reply:
Nice.
Should I make some changes in my patch? Is it already fit to merge?